recipes/m4/all/test_package/conanfile.py

```diff
@@ -1,41 +1,40 @@
-from conans import ConanFile, tools
-from conans.errors import ConanException
-from io import StringIO
 import os
 import textwrap
+from io import StringIO
+
+from conan import ConanFile
+from conan.tools.build import can_run
+from conan.tools.files import save
 
 
 class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
+    generators = "VirtualBuildEnv"
+    test_type = "explicit"
+
+    def build_requirements(self):
+        self.tool_requires(self.tested_reference_str)
 
     @property
     def _m4_input_path(self):
         return os.path.join(self.build_folder, "input.m4")
 
     def build(self):
-        tools.save(self._m4_input_path, textwrap.dedent("""\
+        save(self, self._m4_input_path, textwrap.dedent("""\
             m4_define(NAME1, `Harry, Jr.')
             m4_define(NAME2, `Sally')
             m4_define(MET, `$1 met $2')
             MET(`NAME1', `NAME2')
         """))
 
     def test(self):
-        if hasattr(self, "settings_build"):
-            exe_suffix = ".exe" if self.settings.os == "Windows" else ""
-            m4_bin = os.path.join(self.deps_cpp_info["m4"].rootpath, "bin", "m4" + exe_suffix)
-        else:
-            m4_bin = tools.get_env("M4")
-            if m4_bin is None or not m4_bin.startswith(self.deps_cpp_info["m4"].rootpath):
-                raise ConanException("M4 environment variable not set")
-
-        if not tools.cross_building(self, skip_x64_x86=True):
-            self.run("{} --version".format(m4_bin), run_environment=True)
-            self.run("{} -P {}".format(m4_bin, self._m4_input_path))
+        if can_run(self):
+            self.run("m4 --version", run_environment=True)
+            self.run(f"m4 -P {self._m4_input_path}")
 
-            self.run("m4 -R {0}/frozen.m4f {0}/test.m4".format(self.source_folder), run_environment=True)
+            self.run(f"m4 -R {os.path.join(self.source_folder, 'frozen.m4f')} {os.path.join(self.source_folder, 'test.m4')}", run_environment=True)
```

Comment:
Can you use the `VirtualRunEnv` generator here?

```
self.run(os.path.join(self.cpp.build.bindirs[0], "test_package"), env="conanrun")
```

Reply:
@jwillikers I'm a bit uncertain what that should test then? Since the `m4` package is used as a `tool_requires` and it only needs to be tested as such; Hence the usage of the `VirtualBuildEnv` generator.